File: polygon/src/main/java/com/xattacker/json/JsonUtility.java
package com.xattacker.json;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;

import android.content.Context;

import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.Expose;
import com.google.gson.reflect.TypeToken;

public final class JsonUtility
{
	public static Gson createGson(JsonBuilderVisitor aVisitor)
	{
	   GsonBuilder builder = new GsonBuilder();  
	   
	   if (aVisitor != null)
	   {
	   	aVisitor.onBuilderPrepared(builder);
	   }
		
	   // ignore expose field to serialize
	   builder.setExclusionStrategies
      (
          new ExclusionStrategy()
          {
              @Override
              public boolean shouldSkipField(FieldAttributes f)
              {
                  return f.getAnnotation(Expose.class) != null;
              }

              @Override
              public boolean shouldSkipClass(Class<?> clazz)
              {
                  return false;
              }
          }
      );
	   
	   return builder.create();   
	}
	
	public static Gson createGson()
	{
		return createGson(null);
	}
	
	public static <T> T fromJson(String aJson, TypeToken<T> aToken, JsonBuilderVisitor aVisitor)
	{
		Gson gson = createGson(aVisitor);
		return gson.fromJson(aJson, aToken.getType());
	}
	
	public static <T> T fromJson(String aJson, TypeToken<T> aToken)
	{
		return fromJson(aJson, aToken, null);
	}
	
	public static <T> T fromJson(String aJson, Class<T> aType, JsonBuilderVisitor aVisitor)
	{
		Gson gson = createGson(aVisitor);
		return gson.fromJson(aJson, aType);
	}
	
	public static <T> T fromJson(String aJson, Class<T> aType)
	{
		return fromJson(aJson, aType, null);
	}
	
	public static <T> T fromJsonRes
	(
	Context aContext, 
	int aRes, 
	Class<T> aType, 
	JsonBuilderVisitor aVisitor
	) throws Exception
	{
		T obj = null;
		ByteArrayOutputStream bout = null;
		InputStream in = null;
		
		try
		{
			in = aContext.getResources().openRawResource(aRes);
			bout = new ByteArrayOutputStream();

			byte[] buffer = new byte[256];
			int index = 0;
			
			while ((index = in.read(buffer)) != -1)
			{
				bout.write(buffer, 0, index);
				bout.flush();
			}
			
			buffer = null;

			obj = fromJson 
	      		(
	      		new String(bout.toByteArray()),
	      		aType,
	      		aVisitor
	      	   );
		}
		catch (Exception ex)
		{
			throw ex;
		}
		finally
		{
			if (in != null)
			{
				try
				{
					in.close();
				}
				catch (Exception ex)
				{
				}
				
				in = null;
			}
			
			if (bout != null)
			{
				try
				{
					bout.close();
				}
				catch (Exception ex)
				{
				}
				
				bout = null;
			}
		}

		return obj;
	}
	
	public static <T> T fromJsonRes
	(
	Context aContext, 
	int aRes, 
	Class<T> aType
	) throws Exception
	{
		return fromJsonRes(aContext, aRes, aType, null);
	}
	
	public static <T> T fromJsonRes
	(
	Context aContext, 
	int aRes, 
	TypeToken<T> aToken, 
	JsonBuilderVisitor aVisitor
	) throws Exception
	{
		T obj = null;
		ByteArrayOutputStream bout = null;
		InputStream in = null;
		
		try
		{
			in = aContext.getResources().openRawResource(aRes);
			bout = new ByteArrayOutputStream();

			byte[] buffer = new byte[256];
			int index = 0;
			
			while ((index = in.read(buffer)) != -1)
			{
				bout.write(buffer, 0, index);
				bout.flush();
			}
			
			buffer = null;

			obj = fromJson 
	      		(
	      		new String(bout.toByteArray()),
	      		aToken,
	      		aVisitor
	      	   );
		}
		catch (Exception ex)
		{
			throw ex;
		}
		finally
		{
			if (in != null)
			{
				try
				{
					in.close();
				}
				catch (Exception ex)
				{
				}
				
				in = null;
			}
			
			if (bout != null)
			{
				try
				{
					bout.close();
				}
				catch (Exception ex)
				{
				}
				
				bout = null;
			}
		}

		return obj;
	}
	
	public static <T> T fromJsonRes
	(
	Context aContext, 
	int aRes, 
	TypeToken<T> aToken
	) throws Exception
	{
		return fromJsonRes(aContext, aRes, aToken, null);
	}
}

File: polygon/src/main/java/com/xattacker/json/JsonUtility.kt
package com.xattacker.json

import android.content.Context
import com.google.gson.ExclusionStrategy
import com.google.gson.FieldAttributes
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.annotations.Expose
import com.google.gson.reflect.TypeToken
import java.io.ByteArrayOutputStream
import java.io.InputStream

object JsonUtility
{
    @JvmOverloads
    fun createGson(aVisitor: ((builder: GsonBuilder) -> Unit)? = null): Gson
    {
        val builder = GsonBuilder()
        aVisitor?.invoke(builder)

        // ignore expose field to serialize
        builder.setExclusionStrategies(
            object : ExclusionStrategy
            {
                override fun shouldSkipField(f: FieldAttributes): Boolean
                {
                    return f.getAnnotation<Expose>(Expose::class.java) != null
                }

                override fun shouldSkipClass(clazz: Class<*>): Boolean
                {
                    return false
                }
            }
        )

        return builder.create()
    }

    @JvmOverloads
    fun <T> fromJson(aJson: String, aToken: TypeToken<T>, aVisitor:((builder: GsonBuilder) -> Unit)? = null): T?
    {
        val gson = createGson(aVisitor)
        return gson.fromJson<T>(aJson, aToken.type)
    }

    @JvmOverloads
    fun <T> fromJson(aJson: String, aType: Class<T>, aVisitor: ((builder: GsonBuilder) -> Unit)? = null): T
    {
        val gson = createGson(aVisitor)
        return gson.fromJson(aJson, aType)
    }

    @JvmOverloads
    @Throws(Exception::class)
    fun <T> fromJsonRes(
    context: Context,
    aRes: Int,
    aType: Class<T>,
    aVisitor: ((builder: GsonBuilder) -> Unit)? = null
    ): T?
    {
        var obj: T?
        var bout: ByteArrayOutputStream? = null
        var fin: InputStream? = null

        try
        {
            fin = context.resources.openRawResource(aRes)
            bout = ByteArrayOutputStream()

            val buffer = ByteArray(256)
            var index = 0

            while (true)
            {
                index = fin.read(buffer)
                if (index < 0)
                {
                    break
                }

                bout.write(buffer, 0, index)
                bout.flush()
            }

            obj = fromJson(
                    String(bout.toByteArray()),
                    aType,
                    aVisitor)
        }
        catch (ex: Exception)
        {
            throw ex
        }
        finally
        {
            try
            {
                fin?.close()
            }
            catch (ex: Exception)
            {
            }

            try
            {
                bout?.close()
            }
            catch (ex: Exception)
            {
            }
        }

        return obj
    }

    @JvmOverloads
    @Throws(Exception::class)
    fun <T> fromJsonRes(
    context: Context,
    aRes: Int,
    aToken: TypeToken<T>,
    aVisitor: ((builder: GsonBuilder) -> Unit)?= null
    ): T?
    {
        var obj: T?
        var bout: ByteArrayOutputStream? = null
        var fin: InputStream? = null

        try
        {
            fin = context.resources.openRawResource(aRes)
            bout = ByteArrayOutputStream()

            val buffer = ByteArray(256)
            var index = 0

            while (true)
            {
                index = fin.read(buffer)
                if (index < 0)
                {
                    break
                }

                bout.write(buffer, 0, index)
                bout.flush()
            }

            obj = fromJson(
                    String(bout.toByteArray()),
                    aToken,
                    aVisitor)
        }
        catch (ex: Exception)
        {
            throw ex
        }
        finally
        {
            try
            {
                fin?.close()
            }
            catch (ex: Exception)
            {
            }

            try
            {
                bout?.close()
            }
            catch (ex: Exception)
            {
            }
        }

        return obj
    }
}
